flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/RedundantRankNumberColumnRemoveRule.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical
+
+import org.apache.flink.table.planner.plan.nodes.logical.{FlinkLogicalCalc, FlinkLogicalRank}
+import org.apache.flink.table.planner.plan.utils.InputRefVisitor
+
+import org.apache.calcite.plan.RelOptRule.{any, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rex.RexProgramBuilder
+
+import scala.collection.JavaConversions._
+
+/**
+  * Planner rule that removes the output column of rank number
+  * iff the rank number column is not used by successor calc.
+  */
+class RedundantRankNumberColumnRemoveRule
+  extends RelOptRule(
+    operand(classOf[FlinkLogicalCalc],
+      operand(classOf[FlinkLogicalRank], any())),
+    "RedundantRankNumberColumnRemoveRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val calc: FlinkLogicalCalc = call.rel(0)
+    val rank: FlinkLogicalRank = call.rel(1)
+    val rankNumberColumnIdx = rank.getRowType.getFieldCount - 1
```

Comment:
actually, the result is wrong if `rank.outputRankNumber` is false

Reply:
yes ,but the rule only matches if rank.outputRankNumber is true